### DISCUSSION TITLE
Add hooks & utils to package.json exports?

### DISCUSSION THREAD

**Discussion by ryanburns23** [tailwindlabs/headlessui#2219] [Ideas]
Hello! First off... thanks for the great library!

I am the maintainer of a large UI Kit that uses headless UI under the hood to structure its components. Some of the use-cases in this kit get advanced and we try to re-use logic from headless UI where possible.

For example, we wanted to re-create this [`useScrollLock()`](https://github.dev/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/dialog/dialog.tsx#L94-L95) hook which is found in the dialog component. This scroll lock function depends on a few helpers like [`disposables()`](https://github.dev/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/utils/disposables.ts#L1) & [`isIOS()`](https://github.dev/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/utils/platform.ts#L6)

We do not want to duplicate all the hooks and utils from headlessUI and would rather use them directly by importing them. The files are available in the published build which is great, there may just be one minor tweak needed to make linters like [`eslint-import`](https://github.com/import-js/eslint-plugin-import/) happy.

Let's take the `isIOS` function as an example.

- This function is exported from `@headlessui/react/dist/utils/platform.js`
- Exporting this function like `import { isIOS } from '@headlessui/react/dist/utils/platform';` works, but there is a lint error
- The linting error reported is: `Unable to resolve path to module @headlessui/react/dist/utils/platform. eslint[import/no-unresolved]`

I diagnosed this error stemming from the fact that in the package.json, there are no exports to these utils or hooks folders. 
Currently the package.json has exports listed as:

```json
"exports": {
    "import": "./dist/headlessui.esm.js",
    "require": "./dist/index.cjs",
    "types": "./dist/index.d.ts"
}
```

To resolve this I updated the exports to:
```json
"exports": {
    ".": {
      "import": "./dist/headlessui.esm.js",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts"
    },
    "./dist/hooks/*": {
      "import": "./dist/hooks/*.js",
      "types": "./dist/hooks/*.d.ts"
    },
    "./dist/utils/*": {
      "import": "./dist/utils/*.js",
      "types": "./dist/utils/*.d.ts"
    }
 }
```

![Screen Shot 2023-01-27 at 8 43 47 AM](https://user-images.githubusercontent.com/17202389/215150048-73d922b1-6ef1-4fbf-9595-4d9134fe0c3a.png)
![Screen Shot 2023-01-27 at 8 42 56 AM](https://user-images.githubusercontent.com/17202389/215150073-09b4ccaf-a4c9-4ed6-a9ab-ca1b3835e284.png)

**Are these exports something you would consider adding to the library?**

Thanks

**Comment by ryanburns23** [comment]
In case anyone needs a workaround:

```js
// update-headlessui-package-json.js
const fs = require('fs');
const path = require('path');
const packageJsonPath = path.join(__dirname, '..', 'node_modules', '@headlessui', 'react', 'package.json');
const packageJson = require(packageJsonPath);

packageJson.exports = {
  '.': {
    import: './dist/headlessui.esm.js',
    require: './dist/index.cjs',
    types: './dist/index.d.ts',
  },
  './dist/hooks/*': {
    import: './dist/hooks/*.js',
    types: './dist/hooks/*.d.ts',
  },
  './dist/utils/*': {
    import: './dist/utils/*.js',
    types: './dist/utils/*.d.ts',
  },
};
fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2));

```

```json
"scripts": {
  "postinstall": "node scripts/update-headlessui-package-json.js",
}
```

**Comment by lukasvice** [comment]
This would be a great addition.

We want to use the "useId" hook for our own components. We managed to import it with "import { useId } from "@headlessui/react/dist/hooks/use-id".
However, the problem is that this is only the ESM version. When we want to use it with CJS projects it doesn't work.

If the hooks were exported, there would be a pre-built ESM and CJS version.

**Comment by hdk-pd** [comment]
Adding my voice here. Thanks for the workaround @ryanburns23, would love to be able to remove this from my project though in a future version :)

**Comment by serudda** [comment]
Same from our side... this would be awesome

**Comment by aufdenpunkt** [comment]
Would be great! 🙂
